pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,246 @@
+	return m.SetUpAt(m.GetPath())
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageTypeMemory}},
+}
+
+// readFile reads full path file name and return it as a string
+func readFile(fileName string) string {
+	data, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		fmt.Printf("%s\n", err)
+		return ""
+	}
+	return string(data)
+}
+
+// generateTmpFile writes input parameter 'values' in a temporary file. File descriptor
+// is returned. In case error is returned file descriptor is not meaningful
+func generateTmpFile(outputDir, values string) (*os.File, error) {
+	prefix := "tmpfile"
+	tmpFile, err := ioutil.TempFile(outputDir, prefix)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create tmp file in dir '%s', with prefix '%s'", outputDir, prefix)
+	}
+	err = ioutil.WriteFile(tmpFile.Name(), []byte(values), 0644)
+	if err != nil {
+		return nil, err
+	}
+	return tmpFile, nil
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true,
+func isFileToGenerate(outputDir, fileName, values string) bool {
+	return readFile(path.Join(outputDir, fileName)) != values
+}
+
+// writeInFile writes string in fileName file if string must be updated.
+// It writes 'atomically' writing in a tmp file and re-placing the previous copy
+// via symbol linking recreation. It ends up removing the previous file
+func (m *metadataVolume) writeInFile(fileName, values string) error {
+	if values == "" || fileName == "" {
+		return nil
+	}
+	if !isFileToGenerate(m.GetPath(), fileName, values) {
+		return nil
+	}
+	fd, err := generateTmpFile(m.GetPath(), values)
+	if err != nil {
+		fmt.Printf("%s\n", err)
+		return err
+	}
+
+	// unlink the file
+	finalFilePath := path.Join(m.GetPath(), fileName)
+	actualTarget, e := os.Readlink(finalFilePath)
+	if e != nil {
+		// no link so creates one
+		os.Symlink(fd.Name(), finalFilePath)
+		return nil
+	}
+	e = os.Remove(finalFilePath) // remove the link
+	if e != nil {
+		return e
+	}
+	e = os.Symlink(fd.Name(), finalFilePath) // and create the new one
+	if e != nil {
+		return e
+	}
+	return os.Remove(actualTarget)
+
+}
+
+// formatMap formats map[string]string to a string.
+func formatMap(m map[string]string) string {
+	var l string
+	for key, value := range m {
+		l += key + "=" + value + "\n"
+	}
+	return l
+}
+
+func (m *metadataVolume) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a metadata volume %v for pod %v at %v", m.volName, m.podRef.UID, dir)
+	// Wrap EmptyDir,let it do the setup.
+	wrapped, err := m.plugin.host.NewWrapperBuilder(wrappedVolumeSpec, m.podRef, *m.opts, m.mounter)
+	if err != nil {
+		return err
+	}
+	if err := wrapped.SetUpAt(dir); err != nil {
+		return err
+	}
+
+	for fieldReference, fileName := range m.fieldReferenceFileName {
```

Comment:
I wrote it in this way to handle refresh: labels are (or it will be) modifiable.
From  [your previous comment](https://github.com/GoogleCloudPlatform/kubernetes/pull/5093#issuecomment-93656032) which probably I misinterpreted. 
I thought that `SetUp` was the entry point for refreshing and 'idempotency' didn't mean 'write the same data' but write the same data under the same condition. But again, maybe I missed something.
Ti, may you explain how the 'refresh is going to be triggered' and the relation with idempotency?